core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,695 @@
+        if elem in set_a and elem in set_b:
+            intersection_set.append(elem)
+            set_a.remove(elem)
+            set_b.remove(elem)
+
+    coeff = float(len(intersection_set)) / len(union_set)
+    return coeff
+
+
+def get_program_similarity(fingerprint_a, fingerprint_b):
+    """Find similarity between fingerprint of two programs.
+
+    Args:
+        fingerprint_a: list((int, int)). Fingerprint of first data point.
+        fingerprint_b: list((int, int)). Fingerprint of second data point.
+
+    Returns:
+        float. Similarity between first and second program.
+    """
+    set_a = [h for (h, _) in fingerprint_a]
+    set_b = [h for (h, _) in fingerprint_b]
+    return calc_jaccard_index(set_a, set_b)
+
+
+def generate_top_similars(training_data, top):
+    """Find 'top' nearest neighbours for all programs in dataset.
+
+    Args:
+        training_data: dict. A dictionary containing training training_data.
+            Structure of training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program.
+                        'tokens': list(str). Token list of program.
+                        'k_gram_hashes': list(int). K-gram hash values of
+                            program.
+                        'fingerprint': list(int). Extracted fingerprint of
+                            program.
+                    } for each program in training_data.
+                }
+        top: int. No. of nearest neighbours to be identified.
+
+    Returns:
+        dict. Dictionary containing training data with additional attribute
+        'top_similar' in each data point which contains nearest 'top' no. of
+        data points for each data point. Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                    'fingerprint': list(int). Extracted fingerprint of program.
+                    'top_similar': list(int). Most similar programs to this
+                        program.
+                } for each program in training_data.
+            }
+    """
+    for program_id_1 in training_data:
+        overlaps = []
+        for program_id_2 in training_data:
+            overlap = get_program_similarity(
+                training_data[program_id_1]['fingerprint'],
+                training_data[program_id_2]['fingerprint'])
+            overlaps.append((program_id_2, overlap))
+        training_data[program_id_1]['top_similar'] = sorted(
+            overlaps, key=lambda e: e[1], reverse=True)[:top]
+    return training_data
+
+
+def run_knn(training_data, top_neighbours):
```

Comment:
I don't think so. This is not normal KNN. Here vectors are different and their metric is also different, which is not possible in sklearn. (they don't allow custom metric function, I think)